File: Plugins/AM_ProSightPC_Quant_Plugin/clsAnalysisResourcesProSightQuant.vb
Option Strict On

Imports AnalysisManagerBase

Public Class clsAnalysisResourcesProSightQuant
	Inherits clsAnalysisResources

	Public Const PROSIGHT_PC_RESULT_FILE As String = "ProSightPC_Results.xls"
	Public Const TOOL_DISABLED As Boolean = True

	Public Overrides Function GetResources() As IJobParams.CloseOutType

		If TOOL_DISABLED Then
			' This tool is currently disabled, so just return Success
			Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS
		End If

		' Retrieve the MSAlign_Quant parameter file
		' For example, MSAlign_Quant_Workflow_2012-07-25

		Dim strParamFileStoragePathKeyName As String
		Dim strParamFileStoragePath As String
		strParamFileStoragePathKeyName = clsGlobal.STEPTOOL_PARAMFILESTORAGEPATH_PREFIX & "MSAlign_Quant"

		strParamFileStoragePath = m_mgrParams.GetParam(strParamFileStoragePathKeyName)
		If strParamFileStoragePath Is Nothing OrElse strParamFileStoragePath.Length = 0 Then
			strParamFileStoragePath = "\\gigasax\DMS_Parameter_Files\DeconToolsWorkflows"
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Parameter '" & strParamFileStoragePathKeyName & "' is not defined (obtained using V_Pipeline_Step_Tools_Detail_Report in the Broker DB); will assume: " & strParamFileStoragePath)
		End If

		Dim strParamFileName As String = m_jobParams.GetParam("ProSightQuantParamFile")
		If String.IsNullOrEmpty(strParamFileName) Then
			m_message = clsAnalysisToolRunnerBase.NotifyMissingParameter(m_jobParams, "ProSightQuantParamFile")
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
			Return IJobParams.CloseOutType.CLOSEOUT_NO_PARAM_FILE
		End If

		clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, "Getting data files")

		If Not RetrieveFile(strParamFileName, strParamFileStoragePath) Then
			Return IJobParams.CloseOutType.CLOSEOUT_NO_PARAM_FILE
		End If

		' Retrieve the ProSightPC results for this job
		Dim strProSightPCResultsFile As String
		strProSightPCResultsFile = PROSIGHT_PC_RESULT_FILE
		If Not FindAndRetrieveMiscFiles(strProSightPCResultsFile, False) Then
			'Errors were reported in function call, so just return
			Return IJobParams.CloseOutType.CLOSEOUT_FILE_NOT_FOUND
		End If
		m_jobParams.AddResultFileToSkip(strProSightPCResultsFile)


		' Get the instrument data file
		Dim strRawDataType As String = m_jobParams.GetParam("RawDataType")

		Select Case strRawDataType.ToLower
			Case RAW_DATA_TYPE_DOT_RAW_FILES, RAW_DATA_TYPE_BRUKER_FT_FOLDER
				If RetrieveSpectra(strRawDataType) Then

					If Not MyBase.ProcessMyEMSLDownloadQueue(m_WorkingDir, MyEMSLReader.Downloader.DownloadFolderLayout.FlatNoSubfolders) Then
						Return IJobParams.CloseOutType.CLOSEOUT_FAILED
					End If

					' Confirm that the .Raw or .D folder was actually copied locally
					If strRawDataType.ToLower() = RAW_DATA_TYPE_DOT_RAW_FILES Then
						If Not System.IO.File.Exists(System.IO.Path.Combine(m_WorkingDir, m_DatasetName & DOT_RAW_EXTENSION)) Then
							m_message = "Thermo .Raw file not successfully copied to WorkDir; likely a timeout error"
							clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "clsDtaGenResources.GetResources: " & m_message)
							Return IJobParams.CloseOutType.CLOSEOUT_FAILED
						End If
						m_jobParams.AddResultFileExtensionToSkip(clsAnalysisResources.DOT_RAW_EXTENSION)  'Raw file

					ElseIf strRawDataType.ToLower() = RAW_DATA_TYPE_BRUKER_FT_FOLDER Then
						If Not System.IO.Directory.Exists(System.IO.Path.Combine(m_WorkingDir, m_DatasetName & DOT_D_EXTENSION)) Then
							m_message = "Bruker .D folder not successfully copied to WorkDir; likely a timeout error"
							clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "clsDtaGenResources.GetResources: " & m_message)
							Return IJobParams.CloseOutType.CLOSEOUT_FAILED
						End If
					End If

				Else
					clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "clsDtaGenResources.GetResources: Error occurred retrieving spectra.")
					Return IJobParams.CloseOutType.CLOSEOUT_FAILED
				End If
			Case Else
				m_message = "Dataset type " & strRawDataType & " is not supported"
				clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "clsDtaGenResources.GetResources: " & m_message & "; must be " & RAW_DATA_TYPE_DOT_RAW_FILES & " or " & RAW_DATA_TYPE_BRUKER_FT_FOLDER)
				Return IJobParams.CloseOutType.CLOSEOUT_FAILED
		End Select

		If Not MyBase.ProcessMyEMSLDownloadQueue(m_WorkingDir, MyEMSLReader.Downloader.DownloadFolderLayout.FlatNoSubfolders) Then
			Return IJobParams.CloseOutType.CLOSEOUT_FAILED
		End If

		Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

	End Function


End Class

File: Plugins/AM_ProSightPC_Quant_Plugin/clsAnalysisResourcesProSightQuant.cs
using System.IO;
using AnalysisManagerBase;

namespace AnalysisManagerProSightQuantPlugIn
{

    /// <summary>
    /// Retrieve resources for the ProSight Quant plugin
    /// </summary>
    public class clsAnalysisResourcesProSightQuant : clsAnalysisResources
    {

        /// <summary>
        /// ProSight PC results file name
        /// </summary>
        public const string PROSIGHT_PC_RESULT_FILE = "ProSightPC_Results.xls";

        /// <summary>
        /// Used to track whether this tool is disabled
        /// </summary>
        public const bool TOOL_DISABLED = true;

        /// <summary>
        /// Retrieve required files
        /// </summary>
        /// <returns>Closeout code</returns>
        public override CloseOutType GetResources()
        {
            // Retrieve shared resources, including the JobParameters file from the previous job step
            var result = GetSharedResources();
            if (result != CloseOutType.CLOSEOUT_SUCCESS)
            {
                return result;
            }

            // ReSharper disable once ConditionIsAlwaysTrueOrFalse
            if (TOOL_DISABLED)
            {
                // This tool is currently disabled, so just return Success
                return CloseOutType.CLOSEOUT_SUCCESS;
            }

#pragma warning disable 162

            // Retrieve the MSAlign_Quant parameter file
            // For example, MSAlign_Quant_Workflow_2012-07-25

            string strParamFileStoragePathKeyName = null;
            string strParamFileStoragePath = null;
            strParamFileStoragePathKeyName = clsGlobal.STEPTOOL_PARAMFILESTORAGEPATH_PREFIX + "MSAlign_Quant";

            strParamFileStoragePath = mMgrParams.GetParam(strParamFileStoragePathKeyName);
            if (string.IsNullOrEmpty(strParamFileStoragePath))
            {
                strParamFileStoragePath = @"\\gigasax\DMS_Parameter_Files\DeconToolsWorkflows";
                LogWarning(
                    "Parameter '" + strParamFileStoragePathKeyName +
                    "' is not defined (obtained using V_Pipeline_Step_Tools_Detail_Report in the Broker DB); will assume: " + strParamFileStoragePath);
            }

            var strParamFileName = mJobParams.GetParam("ProSightQuantParamFile");
            if (string.IsNullOrEmpty(strParamFileName))
            {
                mMessage = clsAnalysisToolRunnerBase.NotifyMissingParameter(mJobParams, "ProSightQuantParamFile");
                LogError(mMessage);
                return CloseOutType.CLOSEOUT_NO_PARAM_FILE;
            }

            LogMessage("Getting data files");

            if (!FileSearch.RetrieveFile(strParamFileName, strParamFileStoragePath))
            {
                return CloseOutType.CLOSEOUT_NO_PARAM_FILE;
            }

            // Retrieve the ProSightPC results for this job
            string strProSightPCResultsFile = null;
            strProSightPCResultsFile = PROSIGHT_PC_RESULT_FILE;
            if (!FileSearch.FindAndRetrieveMiscFiles(strProSightPCResultsFile, false))
            {
                // Errors were reported in function call, so just return
                return CloseOutType.CLOSEOUT_FILE_NOT_FOUND;
            }
            mJobParams.AddResultFileToSkip(strProSightPCResultsFile);

            // Get the instrument data file
            var strRawDataType = mJobParams.GetParam("RawDataType");

            switch (strRawDataType.ToLower())
            {
                case RAW_DATA_TYPE_DOT_RAW_FILES:
                case RAW_DATA_TYPE_BRUKER_FT_FOLDER:

                    if (FileSearch.RetrieveSpectra(strRawDataType))
                    {
                        if (!ProcessMyEMSLDownloadQueue(mWorkDir, MyEMSLReader.Downloader.DownloadFolderLayout.FlatNoSubfolders))
                        {
                            return CloseOutType.CLOSEOUT_FAILED;
                        }

                        // Confirm that the .Raw or .D folder was actually copied locally
                        if (strRawDataType.ToLower() == RAW_DATA_TYPE_DOT_RAW_FILES)
                        {
                            if (!File.Exists(Path.Combine(mWorkDir, DatasetName + DOT_RAW_EXTENSION)))
                            {
                                mMessage = "Thermo .Raw file not successfully copied to WorkDir; likely a timeout error";
                                LogError("clsDtaGenResources.GetResources: " + mMessage);
                                return CloseOutType.CLOSEOUT_FAILED;
                            }

                            // Raw file
                            mJobParams.AddResultFileExtensionToSkip(DOT_RAW_EXTENSION);
                        }
                        else if (strRawDataType.ToLower() == RAW_DATA_TYPE_BRUKER_FT_FOLDER)
                        {
                            if (!Directory.Exists(Path.Combine(mWorkDir, DatasetName + DOT_D_EXTENSION)))
                            {
                                mMessage = "Bruker .D folder not successfully copied to WorkDir; likely a timeout error";
                                LogError("clsDtaGenResources.GetResources: " + mMessage);
                                return CloseOutType.CLOSEOUT_FAILED;
                            }
                        }
                    }
                    else
                    {
                        LogDebug("clsDtaGenResources.GetResources: Error occurred retrieving spectra.");
                        return CloseOutType.CLOSEOUT_FAILED;
                    }
                    break;
                default:
                    mMessage = "Dataset type " + strRawDataType + " is not supported";
                    LogError("clsDtaGenResources.GetResources: " + mMessage +
                             "; must be " + RAW_DATA_TYPE_DOT_RAW_FILES + " or " + RAW_DATA_TYPE_BRUKER_FT_FOLDER);
                    return CloseOutType.CLOSEOUT_FAILED;
            }

            if (!ProcessMyEMSLDownloadQueue(mWorkDir, MyEMSLReader.Downloader.DownloadFolderLayout.FlatNoSubfolders))
            {
                return CloseOutType.CLOSEOUT_FAILED;
            }

            return CloseOutType.CLOSEOUT_SUCCESS;

#pragma warning restore 162
        }
    }
}
